llvm/lib/Analysis/ScalarEvolution.cpp

```diff
@@ -15767,7 +15768,7 @@ void ScalarEvolution::LoopGuards::collectFromBlock(
 
     // If we are recursively collecting guards stop after 2
     // predecessors to limit compile-time impact for now.
-    if (Depth > 0 && Terms.size() == 2)
+    if (Depth > 0 && ++NumCollectedConditions == 2)
```

Comment:
Ah I see, the issue is when we have cases where `Terms` already has entries from assumes.

```suggestion
  ++ NumPredecessors;
    if (Depth > 0 && NumVisitedPredecessors  == 2)
```

This aligns better with the comment immediately above. Might also be worth pulling out the increment of the condition to simplify for the reader

Reply:
Hm, I think it is actually counting the number of conditions, due to the early continue above, changed the comment instead (and pulled out the increment) :)